### PR TITLE
Refactor Case Activity to generic Audit streams.

### PR DESCRIPTION
**Pull request by a-bek**
There are 2 parts of the PR:

1) Introduction of `/audit_trail/` app which is the main part of the work:
        Summary - Does not use the `django-activity-streams` as intended out of the box, rather, we will maintain our own model but make use of some utilities `django-activity-streams` provides (Managers and Querysets)

2) Removing `BaseActivity`, `CaseActivity`, completely from the API, replacing all CaseActivity calls and tests the `Audit`

Several TODOs which I added, these include:
     - Audit Trail test coverage
     - Parametrising the serialisation of Audits
     - Compressing the Verbs into a generic set of Verbs, as it is now I have just copied over from CaseActivity

### CODE REVIEW

File: audit_trail/urls.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+app_name = "applications"
```
Comment (Mark-Spiers-kainos) [inline audit_trail/urls.py]: think this app name should be changed

File: cases/views/views.py [hunk 2]
```diff
@@ -234,9 +216,13 @@ def get(self, request, pk):
             advice = self.advice.filter(user__team=team)
             create_grouped_advice(self.case, self.request, advice, TeamAdvice)
             case_advice_contains_refusal(pk)
-            CaseActivity.create(
-                activity_type=CaseActivityType.CREATED_TEAM_ADVICE, case=self.case, user=request.user,
+
+            audit_trail_service.create(
```
Comment (Mark-Spiers-kainos) [inline cases/views/views.py]: when is `audit_trail_service.create` used in comparison to `Audit.objects.create`? See both in this file

File: applications/views/external_locations.py [hunk 3]
```diff
@@ -131,7 +132,38 @@ def post(self, request, application):
...
         removed_sites = SiteOnApplication.objects.filter(application=application)
 
-        set_external_location_case_activity(removed_sites, removed_locations, new_locations, request.user, application)
+        if removed_sites:
+            audit_trail_service.create(
+                actor=request.user,
+                verb=Verb.REMOVED_SITES_FROM_APPLICATION,
+                target=application.get_case() or application,
+                payload={
+                    'sites': [site.site.name + " " + site.site.address.country.name for site in removed_sites],
```
Comment (andytaylorkainos) [inline applications/views/external_locations.py]: not sure if we should separate the site name and the country by a comma. Might be worth asking Anthony because it'll end up on the frontend :)

File: applications/views/goods.py [hunk 4]
```diff
@@ -85,8 +82,14 @@ def post(self, request, application):
...
 
-                set_application_goods_case_activity(
-                    CaseActivityType.ADD_GOOD_TO_APPLICATION, good.description, request.user, application,
+                audit_trail_service.create(
+                    actor=request.user,
+                    verb=Verb.ADDED_GOOD_TO_APPLICATION,
+                    action_object=good,
+                    target=application.get_case() or application,
+                    payload={
+                        'good': good.description
```
Comment (andytaylorkainos) [inline applications/views/goods.py]: im guessing black hasnt complained at you for the single quotes yet? 😅

File: applications/views/sites.py [hunk 5]
```diff
@@ -119,9 +120,50 @@ def post(self, request, application):
...
+            added_sites=new_sites
+        )
 
         removed_sites.delete()
         removed_locations.delete()
 
         return JsonResponse(data={"sites": response_data}, status=status.HTTP_201_CREATED)
+
+
+def set_activity(user, application, removed_locations, removed_sites, added_sites):
```
Comment (andytaylorkainos) [inline applications/views/sites.py]: prepend this method with `_`

File: audit_trail/apps.py [hunk 6]
```diff
@@ -0,0 +1,6 @@
+from django.apps import AppConfig
+
+
+class AuditTrailConfig(AppConfig):
```
Comment (andytaylorkainos) [inline audit_trail/apps.py]: Dont think were using these files, can you remove it please

File: audit_trail/models.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    )
+    action_object_object_id = models.CharField(
+        max_length=255, blank=True, null=True, db_index=True
+    )
+    action_object = GenericForeignKey(
+        'action_object_content_type',
+        'action_object_object_id'
+    )
+
+    payload = JSONField()
```
Comment (andytaylorkainos) [inline audit_trail/models.py]: we cant use these fields, it'll lock us into using Postgres :(

File: applications/libraries/document_helpers.py [hunk 8]
```diff
@@ -52,8 +49,11 @@ def upload_application_document(application, data, user):
         return JsonResponse({"errors": serializer.errors}, status=status.HTTP_400_BAD_REQUEST)
     serializer.save()
 
-    set_application_document_case_activity(
-        CaseActivityType.UPLOAD_APPLICATION_DOCUMENT, data.get("name"), user, application,
+    audit_trail_service.create(
+        actor=user,
+        verb=Verb.UPLOADED_DOCUMENT,
```
Comment (BenAndrew-Kainos) [inline applications/libraries/document_helpers.py]: Are these verbs defined in lite_content?

File: applications/libraries/document_helpers.py [hunk 9]
```diff
@@ -105,8 +112,15 @@ def delete_party_document(party, application, user):
...
         document.delete()
 
-        set_party_document_case_activity(
-            CaseActivityType.DELETE_PARTY_DOCUMENT, document.name, party.type, party.name, user, application,
+        audit_trail_service.create(
+            actor=user,
+            verb=Verb.DELETED_DOCUMENT,
+            target=application.get_case() or application,
+            payload={
+                "party_type": party.type.replace("_", " "),
```
Comment (BenAndrew-Kainos) [inline applications/libraries/document_helpers.py]: party.type should have a value field if it's being used for display purposes (i.e. use a keyValue field)

File: applications/views/parties.py [hunk 10]
```diff
@@ -50,16 +50,24 @@ def post(self, request, application):
...
-                previous_end_user.type,
-                previous_end_user.name,
-                request.user,
-                application,
+            audit_trail_service.create(
+                actor=request.user,
+                verb=Verb.REMOVED_PARTY,
+                target=application.get_case() or application,
+                payload={
+                    "party_type": previous_end_user.type.replace("_", " "),
```
Comment (BenAndrew-Kainos) [inline applications/views/parties.py]: Same as above

File: applications/views/parties.py [hunk 11]
```diff
@@ -50,16 +50,24 @@ def post(self, request, application):
...
             )
 
-        set_party_case_activity(
-            CaseActivityType.ADD_PARTY, new_end_user.type, new_end_user.name, request.user, application,
+        audit_trail_service.create(
+            actor=request.user,
+            verb=Verb.ADDED_PARTY,
+            target=application.get_case() or application,
+            payload={
+                "party_type": new_end_user.type.replace("_", " "),
```
Comment (BenAndrew-Kainos) [inline applications/views/parties.py]: Same as above

File: audit_trail/constants.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
+from enum import Enum
+
+
+# TODO: Compress verbs.
+class Verb(Enum):
+    ADDED_PARTY = 'added party'
```
Comment (BenAndrew-Kainos) [inline audit_trail/constants.py]: Please define this in lite_content instead

File: audit_trail/managers.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
+from actstream.gfk import GFKQuerySet, GFKManager
+
+
+
+class AuditQuerySet(GFKQuerySet):
+    pass
```
Comment (BenAndrew-Kainos) [inline audit_trail/managers.py]: Please remove unused code

File: audit_trail/managers.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    # def delete(self):
+    #     raise NotImplementedError('Delete not allowed for Audit trail.')
+
+
+class AuditManager(GFKManager):
+    def get_query_set(self):
+        return AuditQuerySet(self.model)
+    get_queryset = get_query_set
+    #
+    # def delete(self):
```
Comment (BenAndrew-Kainos) [inline audit_trail/managers.py]: Please remove unused code

File: audit_trail/managers.py [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+class AuditQuerySet(GFKQuerySet):
+    pass
+    # def delete(self):
+    #     raise NotImplementedError('Delete not allowed for Audit trail.')
+
+
+class AuditManager(GFKManager):
+    def get_query_set(self):
+        return AuditQuerySet(self.model)
+    get_queryset = get_query_set
```
Comment (BenAndrew-Kainos) [inline audit_trail/managers.py]: Why is this necessary (is get_queryset in someway significant to the class)?

File: audit_trail/managers.py [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+    #
+    # def delete(self):
+    #     raise NotImplementedError('Delete not allowed for Audit trail.')
+
+    def create(self, *args, **kwargs):
+        # TODO: decouple notifications and audit (signals?)
+        audit = super(AuditManager, self).create(*args, **kwargs)
+        from users.models import ExporterUser
```
Comment (BenAndrew-Kainos) [inline audit_trail/managers.py]: Why do you need the imports defined here?

File: audit_trail/managers.py [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+
+
+class AuditQuerySet(GFKQuerySet):
+    pass
+    # def delete(self):
+    #     raise NotImplementedError('Delete not allowed for Audit trail.')
+
+
+class AuditManager(GFKManager):
```
Comment (BenAndrew-Kainos) [inline audit_trail/managers.py]: Please add docstring

File: audit_trail/models.py [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+from django.db import models
+from django.utils import timesince
+from django.utils.translation import ugettext as _
+
+from audit_trail.managers import AuditManager
+from common.models import TimestampableModel
+
+
+class Audit(TimestampableModel):
+    """
```
Comment (BenAndrew-Kainos) [inline audit_trail/models.py]: Please fill in docstring

File: audit_trail/models.py [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+    """
+    actor_content_type = models.ForeignKey(
+        ContentType, related_name='actor', null=True,
+        on_delete=models.SET_NULL, db_index=True,
+    )
+    actor_object_id = models.CharField(max_length=255, db_index=True, null=True)
+    actor = GenericForeignKey('actor_content_type', 'actor_object_id')
```
Comment (BenAndrew-Kainos) [inline audit_trail/models.py]: Could you define these actor properties as a separate model (and link as foreignkey) for clarity?

File: audit_trail/models.py [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+        related_name='target',
+        on_delete=models.SET_NULL, db_index=True
+    )
+    target_object_id = models.CharField(
+        max_length=255, blank=True, null=True, db_index=True
+    )
+    target = GenericForeignKey(
+        'target_content_type',
+        'target_object_id'
+    )
```
Comment (BenAndrew-Kainos) [inline audit_trail/models.py]: Same as above

File: audit_trail/models.py [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+        related_name='action_object',
+        on_delete=models.SET_NULL, db_index=True
+    )
+    action_object_object_id = models.CharField(
+        max_length=255, blank=True, null=True, db_index=True
+    )
+    action_object = GenericForeignKey(
+        'action_object_content_type',
+        'action_object_object_id'
+    )
```
Comment (BenAndrew-Kainos) [inline audit_trail/models.py]: Same as above

File: audit_trail/models.py [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+
+    payload = JSONField()
+
+    objects = AuditManager()
+
+    class Meta:
+        ordering = ('-created_at',)
+
+    def __str__(self):
+        ctx = {
```
Comment (BenAndrew-Kainos) [inline audit_trail/models.py]: could you make this variable name clearer?

File: audit_trail/schema.py [hunk 23]
```diff
@@ -0,0 +1,67 @@
+from functools import wraps
+
+from audit_trail.constants import Verb
+from cases.models import Case, CaseNote
+from users.models import GovUser
+
+from collections import namedtuple
+
+
+class Schema(namedtuple('Schema', 'actor verb action_object target payload')):
```
Comment (BenAndrew-Kainos) [inline audit_trail/schema.py]: Could you add a docstring to explain what this does

File: audit_trail/schema.py [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+                    kwargs[field] = None
+
+            kwarg_schema = {key: type(value) for key, value in {**kwargs, **missing_kwargs}.items()}
+            return Schema(**kwarg_schema)
+
+        except TypeError:
+            raise TypeError(f'INVALID KWARGS: {kwargs}')
+
+
+class Registry:
```
Comment (BenAndrew-Kainos) [inline audit_trail/schema.py]: Could you add a docstring to explain

File: audit_trail/schema.py [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+    def validate(self, schema):
+        return schema in self.__registry
+
+
+registry = Registry()
+
+# Case activity
+for schema in [
+    Schema(actor=GovUser, verb=Verb.ADDED_NOTE, action_object=CaseNote, target=Case, payload=dict),
+    Schema(actor=GovUser, verb=Verb.REMOVED_NOTE, action_object=CaseNote, target=Case, payload=dict),
```
Comment (BenAndrew-Kainos) [inline audit_trail/schema.py]: Could you define this as a constant list at the top (perhaps as a dict containing verb & action_object)

File: audit_trail/serializers.py [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+            # is used to reference the Queue names.
+            queues = sorted([q for q in instance.payload['queues']])
+            return f"{instance.verb}: {', '.join(queues)}"
+
+        if verb in [Verb.ADDED_FLAGS, Verb.REMOVED_FLAGS]:
+            from cases.models import Case
+            if isinstance(instance.action_object, Case):
+                return f"{instance.verb} on case: {instance.action_object.id}"
+            else:
+                return f"{instance.verb} on good: {instance.action_object.description}"
```
Comment (BenAndrew-Kainos) [inline audit_trail/serializers.py]: Could all these verb strings be defined elsewhere and strings handled as a function like:
```
def get_audit_message(verb, instance):
    return audit_messages[verb]
```

File: audit_trail/tests/test_audit_trail.py [hunk 27]
```diff
@@ -0,0 +1,11 @@
...
+
+
+class CasesAuditTrail(DataTestClient):
+    # TODO: test schema and creation
+    def setUp(self):
+        super().setUp()
+        self.draft = self.create_open_application(self.organisation)
+
+    def test_cases_audit_trail(self):
+        assert 1
```
Comment (BenAndrew-Kainos) [inline audit_trail/tests/test_audit_trail.py]: If only tests were this easy 😄

File: audit_trail/urls.py [hunk 28]
```diff
@@ -0,0 +1,4 @@
+app_name = "audit_trail"
+
+urlpatterns = [
```
Comment (BenAndrew-Kainos) [inline audit_trail/urls.py]: Could you drop this file as it is unused

File: audit_trail/models.py (hunk 7 above)
Comment (a-bek) [reply audit_trail/models.py]: All environments are postgres? Is everyone's local not using postgres?

File: cases/service.py [hunk 29]
```diff
@@ -0,0 +1,32 @@
+from audit_trail import service as audit_trail_service
+from audit_trail.constants import Verb
+from cases.models import CaseAssignment
+
+
+def update_case_queues(user, case, queues):
```
Comment (BenAndrew-Kainos) [inline cases/service.py]: please add docstring

File: cases/service.py [hunk 30]
```diff
@@ -0,0 +1,32 @@
...
+    update_queues = set([q.name for q in queues])
+    initial_queues = set(case.queues.values_list("name", flat=True))
+    remove_queues = initial_queues - update_queues
```
Comment (BenAndrew-Kainos) [inline cases/service.py]: rename to `removed_queues` maybe

File: cases/views/activity.py [hunk 31]
```diff
@@ -16,14 +17,16 @@ class Activity(APIView):
...
         case = get_case(pk)
-        activity = get_case_activity(case)
-        activity.extend(convert_case_notes_to_activity(get_case_notes_from_case(case, False)))
 
-        # Sort the activity based on date (newest first)
-        activity.sort(key=lambda x: x.created_at, reverse=True)
+        qs = Audit.objects.all()
+
+        q2 = Q(action_object_object_id=case.id, action_object_content_type=ContentType.objects.get_for_model(case))
+        q1 = Q(target_object_id=case.id, target_content_type=ContentType.objects.get_for_model(case))
```
Comment (BenAndrew-Kainos) [inline cases/views/activity.py]: Please rename `qs` `q2` and `q1` to explain what they are

File: common/models.py [hunk 32]
```diff
@@ -3,8 +3,14 @@
 from model_utils.fields import AutoCreatedField, AutoLastModifiedField
 
 
-class TimestampedModel(models.Model):
+class CreatedAt(models.Model):
```
Comment (BenAndrew-Kainos) [inline common/models.py]: Why do we have a model with a single field (can the created_at field not just be added where it is used)?

File: conf/settings.py [hunk 33]
```diff
@@ -182,3 +183,14 @@
 # Security settings
 
 SECURE_BROWSER_XSS_FILTER = True
+
+
+# Django Activity Stream settings
+
+# ACTSTREAM_SETTINGS = {
+#     'FETCH_RELATIONS': True,
```
Comment (BenAndrew-Kainos) [inline conf/settings.py]: Please remove unused code

File: queries/control_list_classifications/tests.py [hunk 34]
```diff
@@ -86,10 +88,12 @@ def test_respond_to_control_list_classification_query_without_updating_control_c
         self.assertEqual(self.query.good.is_good_controlled, str(self.data["is_good_controlled"]))
         self.assertEqual(self.query.good.status, GoodStatus.VERIFIED)
 
-        # Check that only the response activity item has been added
-        case_activities = get_case_activity(self.query.case.get())
-        self.assertEqual(len(case_activities), 1)
-        self.assertEqual(case_activities[0].type, "clc_response")
+        # Check that an audit item has been added
+        qs = Audit.objects.filter(
```
Comment (BenAndrew-Kainos) [inline queries/control_list_classifications/tests.py]: please rename this variable

File: audit_trail/constants.py (hunk 12 above)
Comment (a-bek) [reply audit_trail/constants.py]: I will add them to lite_content when I have finished the `TODO`. Also, all other Enums in api do not use lite_content, eg `parties/enums.py`

File: audit_trail/tests/test_audit_trail.py (hunk 27 above)
Comment (a-bek) [reply audit_trail/tests/test_audit_trail.py]: I know, I figured the tests I updated covering `CaseActivity` will be sufficient for now. Once i do the TODOs, i'll add the tests (seperate PR).

File: audit_trail/serializers.py (hunk 26 above)
Comment (a-bek) [reply audit_trail/serializers.py]: Yes, but not part of this change `TODO: Parametrize Serializer` 

This is just MVP to refactor CaseActivity. Unit tests pass, e2e pass. I'm going to add another PR after I finish my story to clean up. The way I handle this will depend on the final definition of the Verbs. Right now the Verbs are just a straight mapping from `CaseActivity`

File: common/models.py (hunk 32 above)
Comment (a-bek) [reply common/models.py]: These standardise how we timestamp the models and should be used on all our models that require timestamps. I seperated `CreatedAt` and `update_at` because I assume not all models necessarily want `updated_at`.

File: applications/views/parties.py (hunk 11 above)
Comment (a-bek) [reply applications/views/parties.py]: For this mvp of audit I've just mapped how `CaseActivity` retrieved its data to minimize the changes. What really should be done is `class PartyType` should be changed to `class PartyType(Enum):` but this PR is already quite big. Im going to do a second round of changes once I get LT-1489 out of the way

File: applications/views/goods.py (hunk 4 above)
Comment (a-bek) [reply applications/views/goods.py]: 😢

File: applications/views/external_locations.py (hunk 3 above)
Comment (a-bek) [reply applications/views/external_locations.py]: This is how CaseActivity generated the `sites`. I think any design/data related things should be outside the scope of this PR.

**Comment by efahy** [comment]
In UAT I generated a case doc and got this activity text: "Eugene Fahy generated the case document 00001205155812-8295b.pdf from template 00001205155812." on your branch I generated one and got this activity text: "Eugene Fahy generate case document:". They should be the same, right?

**Comment by efahy** [comment]
I get a crash when trying to move a case onto a user created queue on your branch, which I don't get in UAT.  It's an API error but doesn't appear to be related to activity stream code.  May be due master changes not merged in yet??

  **File "/Users/eugene/.pyenv/versions/3.7.4/lib/python3.7/contextlib.py", line 74, in inner
    return func(*args, **kwds)
  File "/Users/eugene/dit/lite-api/cases/views/views.py", line 66, in put
    user=request.user, case=Case.objects.get_obj(case), queues=serializer.validated_data["queues"]
  File "/Users/eugene/dit/lite-api/cases/service.py", line 11, in update_case_queues
    initial_queues = set(case.queues.values_list("name", flat=True))
AttributeError: 'NoneType' object has no attribute 'queues'**

**Comment by efahy** [comment]
No audit trail created when I changed the status on a case.  There is in UAT

**Comment by efahy** [comment]
API error when I try to attach a document to a case.  Works in UAT:

  File "/Users/eugene/.local/share/virtualenvs/lite-api-Gw9bxCGa-/Users/eugene/.pyenv/shims/python/lib/python3.7/site-packages/rest_framework/views.py", line 492, in dispatch
    response = handler(request, *args, **kwargs)
  File "/Users/eugene/.pyenv/versions/3.7.4/lib/python3.7/contextlib.py", line 74, in inner
    return func(*args, **kwds)
  File "/Users/eugene/dit/lite-api/cases/views/views.py", line 108, in post
    actor=request.user, verb=Verb.UploadDocument, target=case, payload={"file_name": document["name"]}
  File "/Users/eugene/.local/share/virtualenvs/lite-api-Gw9bxCGa-/Users/eugene/.pyenv/shims/python/lib/python3.7/enum.py", line 349, in __getattr__
    raise AttributeError(name) from None
AttributeError: UploadDocument

**Comment by efahy** [comment]
API error when I try to attach a case flag to a case:

  File "/Users/eugene/.local/share/virtualenvs/lite-api-Gw9bxCGa-/Users/eugene/.pyenv/shims/python/lib/python3.7/site-packages/rest_framework/serializers.py", line 530, in to_representation
    ret[field.field_name] = field.to_representation(attribute)
  File "/Users/eugene/.local/share/virtualenvs/lite-api-Gw9bxCGa-/Users/eugene/.pyenv/shims/python/lib/python3.7/site-packages/rest_framework/fields.py", line 1889, in to_representation
    return method(value)
  File "/Users/eugene/dit/lite-api/audit_trail/serializers.py", line 42, in get_text
    return f"{instance.verb} on good: {instance.action_object.description}"
AttributeError: 'NoneType' object has no attribute 'description'

**Comment by efahy** [comment]
On UAT when I submitted an open application as an exporter, it added these activities, none of which appeared on your branch for a similar new open application:
- wefg sdfg added good type: <good name>
- wefg sdfg added sites: <site names>
- wefg sdfg added countries: <country list>
- wefg sdfg updated the status to submitted.

**Comment by efahy** [comment]
On your branch, when I removed countries as an exporter on a minor edit I didn't get an activity.  On UAT, i got:
wefg sdfg removed countries: Ajman, Akrotiri, ......

File: audit_trail/models.py (hunk 7 above)
Comment (andytaylorkainos) [reply audit_trail/models.py]: we've got a requirement to be database agnostic unfortunately

**Comment by efahy** [comment]
When I withdraw an application in UAT, i get the following activity but no activity on your branch:

wefg sdfg updated the status to withdrawn.

**Comment by efahy** [comment]
On UAT when I went through giving advice, combining advice, finalising advice and granting a license, I got these activities:
Eugene Fahy updated the status to finalised.
Eugene Fahy created final advice.
Eugene Fahy created team advice.

On your branch, I only got:
Eugene Fahy created final advice:

File: cases/service.py [hunk 35]
```diff
@@ -0,0 +1,35 @@
...
+    """
+    update_queues = set([q.name for q in queues])
+    initial_queues = set(case.queues.values_list("name", flat=True))
+    removed_queues = initial_queues - update_queues
+    new_queues = update_queues - initial_queues
+
+    if removed_queues:
+        CaseAssignment.objects.filter(queue__name__in=removed_queues).delete()
+
+        audit_trail_service.create(
```
Comment (efahy) [inline cases/service.py]: As a general point, there seem to be a lot of instances where you don't set the action object.  In this case, for example, shouldn't the queue be the action object?  Similarly for audits involving parties - I would hope that the action object would be the party being acted on

File: audit_trail/models.py [hunk 36]
```diff
@@ -0,0 +1,75 @@
...
+        'target_object_id'
+    )
+
+    action_object_content_type = models.ForeignKey(
+        ContentType, blank=True, null=True,
+        related_name='action_object',
+        on_delete=models.SET_NULL, db_index=True
+    )
+    action_object_object_id = models.CharField(
+        max_length=255, blank=True, null=True, db_index=True
```
[the remark this answers is not in the document]
Comment (a-bek) [reply audit_trail/models.py]: Audits will not cascade, it can't be set on `object_id` because thats not a foreign key.

File: audit_trail/models.py [hunk 37]
```diff
@@ -0,0 +1,75 @@
...
+    verb = models.CharField(max_length=255, db_index=True)
+    description = models.TextField(blank=True, null=True)
+
+    target_content_type = models.ForeignKey(
+        ContentType, blank=True, null=True,
+        related_name='target',
+        on_delete=models.SET_NULL, db_index=True
+    )
+    target_object_id = models.CharField(
+        max_length=255, blank=True, null=True, db_index=True
```
[the remark this answers is not in the document]
Comment (a-bek) [reply audit_trail/models.py]: We need these to be null in the case where the object itself is deleted so the audit does not cascade.

File: audit_trail/models.py (hunk 19 above)
Comment (a-bek) [reply audit_trail/models.py]: Im not sure i understand?

File: audit_trail/schema.py (hunk 24 above)
Comment (efahy) [reply audit_trail/schema.py]: No docstring comment added - unresolving

File: audit_trail/models.py (hunk 7 above)
Comment (a-bek) [reply audit_trail/models.py]: done

File: applications/libraries/document_helpers.py [hunk 38]
```diff
@@ -106,8 +114,11 @@ def delete_party_document(party, application, user):
         document.delete_s3()
         document.delete()
 
-        set_party_document_case_activity(
-            CaseActivityType.DELETE_PARTY_DOCUMENT, document.name, party.type, party.name, user, application,
+        audit_trail_service.create(
+            actor=user,
+            verb=AuditType.DELETE_PARTY_DOCUMENT,
+            target=application.get_case(),
+            payload={"party_type": party.type.replace("_", " "), "party_name": party.name, "file_name": document.name},
```
Comment (BenAndrew-Kainos) [inline applications/libraries/document_helpers.py]: As mentioned before this should Use an Enum of some kind to match the type to a display value (even if at the moment the value if the key with _ removed)

File: applications/views/parties.py [hunk 39]
```diff
@@ -45,21 +46,21 @@ def post(self, request, application):
...
-                CaseActivityType.REMOVE_PARTY,
-                previous_end_user.type,
-                previous_end_user.name,
-                request.user,
-                application,
+            audit_trail_service.create(
+                actor=request.user,
+                verb=AuditType.REMOVE_PARTY,
+                target=case,
+                payload={"party_type": previous_end_user.type.replace("_", " "), "party_name": previous_end_user.name,},
```
Comment (BenAndrew-Kainos) [inline applications/views/parties.py]: Same as above

File: applications/views/parties.py [hunk 40]
```diff
@@ -45,21 +46,21 @@ def post(self, request, application):
...
             )
 
-        set_party_case_activity(
-            CaseActivityType.ADD_PARTY, new_end_user.type, new_end_user.name, request.user, application,
+        audit_trail_service.create(
+            actor=request.user,
+            verb=AuditType.ADD_PARTY,
+            target=case,
+            payload={"party_type": new_end_user.type.replace("_", " "), "party_name": new_end_user.name,},
```
Comment (BenAndrew-Kainos) [inline applications/views/parties.py]: Same as above

File: applications/views/sites.py [hunk 41]
```diff
@@ -125,9 +126,47 @@ def post(self, request, application):
...
+
+
+def _set_activity(user, application, removed_locations, removed_sites, added_sites):
+    case = application.get_case()
+    if removed_sites:
+        audit_trail_service.create(
+            actor=user,
+            verb=AuditType.REMOVED_SITES_FROM_APPLICATION,
+            target=case,
+            payload={"sites": [site.site.name + " " + site.site.address.country.name for site in removed_sites],},
```
Comment (BenAndrew-Kainos) [inline applications/views/sites.py]: Could this concatenation of name & country.name be moved to a separate method (seems to be used in a lot of places, and will be difficult to update if this format changes)

File: audit_trail/managers.py [hunk 42]
```diff
@@ -0,0 +1,28 @@
+from actstream.gfk import GFKQuerySet, GFKManager
+
+from users.models import ExporterUser
+from users.models import GovUser
+
+
+class AuditQuerySet(GFKQuerySet):
```
Comment (BenAndrew-Kainos) [inline audit_trail/managers.py]: Still unused?

File: audit_trail/models.py [hunk 43]
```diff
@@ -0,0 +1,69 @@
...
+        context = {
+            "actor": self.actor,
+            "verb": self.verb,
+            "action_object": self.action_object,
+            "target": self.target,
+            "age": self.age(),
+        }
+        if self.target:
+            if self.action_object:
+                return _("%(actor)s %(verb)s %(action_object)s on %(target)s %(age)s ago") % context
```
Comment (BenAndrew-Kainos) [inline audit_trail/models.py]: Could these string patterns be defined in constants somewhere

File: audit_trail/schema.py [hunk 44]
```diff
@@ -0,0 +1,73 @@
...
+
+
+check_kwargs = registry.check_kwargs
+
+
+def validate_kwargs(func):
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        # TODO: compress Verb and activate schema validations.
+        # check_kwargs(**kwargs)
```
Comment (BenAndrew-Kainos) [inline audit_trail/schema.py]: Please remove comment

File: audit_trail/serializers.py [hunk 45]
```diff
@@ -0,0 +1,43 @@
...
+            try:
+                if isinstance(payload[key], list):
+                    payload[key] = ", ".join(payload[key])
+            except KeyError as e:
+                print(f"Audit serialization exception skipped: {e}")
+
+        return verb.format(payload)
+
+    def get_additional_text(self, instance):
+        return instance.payload.get("additional_text", "")
```
Comment (BenAndrew-Kainos) [inline audit_trail/serializers.py]: If this methodField is just getting `additional_text` do you need it? Could you not just point it to the additonal_text field

File: cases/managers.py [hunk 46]
```diff
@@ -103,3 +103,44 @@ def search(
...
+            return StandardApplication.objects.get(baseapplication_ptr__case_ptr=case)
+        except StandardApplication.DoesNotExist:
+            pass
+
+        from applications.models import OpenApplication
+
+        try:
+            return OpenApplication.objects.get(baseapplication_ptr__case_ptr=case)
+        except OpenApplication.DoesNotExist:
+            pass
```
Comment (BenAndrew-Kainos) [inline cases/managers.py]: his method should raise an exception if the case is not found in Standard or Open applications

File: cases/managers.py [hunk 47]
```diff
@@ -103,3 +103,44 @@ def search(
...
+            return ControlListClassificationQuery.objects.get(query_ptr__case_ptr=case)
+        except ControlListClassificationQuery.DoesNotExist:
+            pass
+
+        from queries.end_user_advisories.models import EndUserAdvisoryQuery
+
+        try:
+            return EndUserAdvisoryQuery.objects.get(query_ptr__case_ptr=case)
+        except EndUserAdvisoryQuery.DoesNotExist:
+            pass
```
Comment (BenAndrew-Kainos) [inline cases/managers.py]: Same as above

File: cases/tests/tests_create_final_advice.py [hunk 48]
```diff
@@ -2,6 +2,7 @@
 from parameterized import parameterized
 from rest_framework import status
 
+from audit_trail.models import Audit
```
Comment (BenAndrew-Kainos) [inline cases/tests/tests_create_final_advice.py]: This import has been added but no other code has been updated??

File: users/models.py [hunk 49]
```diff
@@ -144,18 +144,22 @@ def unassign_from_cases(self):
         self.case_assignments.clear()
 
     # pylint: disable=W0221
-    def send_notification(self, case_activity=None):
-        # getting models due to circular imports
-        Notification = get_model("cases.Notification")
+    def send_notification(self, audit=None):
```
Comment (BenAndrew-Kainos) [inline users/models.py]: This method shouldn't have an optional argument as you throw an exception if it is not passed.

File: audit_trail/serializers.py (hunk 45 above)
Comment (a-bek) [reply audit_trail/serializers.py]: There is no additional text field on the model instance

File: users/models.py (hunk 49 above)
Comment (a-bek) [reply users/models.py]: As with the `payload`, this is old behaviour, outside scope of this refactor. 

Otherwise, we can continually refactor the whole project in this PR before it ever gets merged.

File: audit_trail/managers.py (hunk 42 above)
Comment (a-bek) [reply audit_trail/managers.py]: What is unused?

File: audit_trail/schema.py (hunk 44 above)
Comment (a-bek) [reply audit_trail/schema.py]: It's a TODO comment.

File: audit_trail/managers.py (hunk 42 above)
Comment (BenAndrew-Kainos) [reply audit_trail/managers.py]: This class just does a pass. Why is it needed?

File: audit_trail/schema.py (hunk 44 above)
Comment (BenAndrew-Kainos) [reply audit_trail/schema.py]: I was referring to the line under it

File: audit_trail/serializers.py (hunk 45 above)
Comment (BenAndrew-Kainos) [reply audit_trail/serializers.py]: OK. Should there not be one?

File: users/models.py (hunk 49 above)
Comment (BenAndrew-Kainos) [reply users/models.py]: I'll leave it as Andy is going to work on it but it is very simple

File: applications/libraries/document_helpers.py (hunk 38 above)
Comment (BenAndrew-Kainos) [reply applications/libraries/document_helpers.py]: As discussed we'll leave this to a later refactor (please leave a TODO)

File: audit_trail/schema.py (hunk 44 above)
Comment (a-bek) [reply audit_trail/schema.py]: The TODO is to uncomment that that line

File: audit_trail/serializers.py (hunk 45 above)
Comment (a-bek) [reply audit_trail/serializers.py]: This serialiser serializes the new audit trail to be compatible with the frontend for the old caseactivity. `additional_text` is part of the old caseactivity

File: audit_trail/managers.py (hunk 42 above)
Comment (a-bek) [reply audit_trail/managers.py]: `AuditQuerySet` is used in the manager. Subclassing it here just lets us customize it if we need going forward.

File: audit_trail/serializers.py (hunk 45 above)
Comment (a-bek) [reply audit_trail/serializers.py]: No, this serializer ensures nothing breaks between frontend and the refactored API changes.

File: audit_trail/models.py (hunk 43 above)
Comment (a-bek) [reply audit_trail/models.py]: `__str__` is subject to change so I don't think it's worth defining it somewhere yet.

Comment (BenAndrew-Kainos) [reply audit_trail/models.py]: Disagree but I'll approve
